README.md
# BitcoinPaymentURI
[![Release](https://jitpack.io/v/SandroMachado/BitcoinPaymentURI.svg)](https://jitpack.io/#SandroMachado/BitcoinPaymentURI)
[![Build Status](https://travis-ci.org/SandroMachado/BitcoinPaymentURI.svg?branch=master)](https://travis-ci.org/SandroMachado/BitcoinPaymentURI)

BitcoinPaymentURI is an open source library to handle the Bitcoin payment URI based on the [BIT 21](https://github.com/bitcoin/bips/blob/master/bip-0021.mediawiki). The purpose of this library is to provide a simplier way to the developers to integrate in their applications support for this URI Scheme  to easily make payments.

# Gradle Dependency

## Repository

First, add the following to your app's `build.gradle` file:

```Gradle
repositories {
    maven { url "https://jitpack.io" }
}
```

Them include the openalpr-android dependency:

```gradle
dependencies {

    // ... other dependencies here.    	
    compile 'com.github.SandroMachado:BitcoinPaymentURI:1.0.3'
}
```

# Usage

## Code

Parse the URI `bitcoin:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W?amount=50&label=Luke-Jr&message=Donation%20for%20project%20xyz`.

```Java
BitcoinPaymentURI bitcoinPaymentURI = BitcoinPaymentURI.parse("bitcoin:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W?amount=50&label=Luke-Jr&message=Donation%20for%20project%20xyz");

bitcoinPaymentURI.getAddress(); \\ 175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W
bitcoinPaymentURI.getAmount(); \\ 50
bitcoinPaymentURI.getLabel(); \\ "Luke-Jr"
bitcoinPaymentURI.getMessage(); \\ "Donation for project xyz"
bitcoinPaymentURI.getParameters().size(); \\ 0
```

Generate the following URI `bitcoin:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W?message=Donation%20for%20project%20xyz&amount=50.0&req-fiz=biz&foo=bar&label=Luke-Jr`

```Java
BitcoinPaymentURI bitcoinPaymentURI = new BitcoinPaymentURI.Builder()
	.address("175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W")
	.amount(50.0)
	.label("Luke-Jr")
	.message("Donation for project xyz")
	.parameter("foo", "bar")
	.requiredParameter("fiz", "biz")
	.build();
```


CHANGELOG.md
# Changelog

## [1.0.3](https://github.com/SandroMachado/BitcoinPaymentURI/tree/1.0.3) (2016-05-04)
[Full Changelog](https://github.com/SandroMachado/BitcoinPaymentURI/compare/1.0.2...1.0.3)

**Closed issues:**

- Wrong parsing for URI with parameters that start with `.\*:` [\#5](https://github.com/SandroMachado/BitcoinPaymentURI/issues/5)

**Merged pull requests:**

- Fix parse for URI with parameters with `.\*:` [\#6](https://github.com/SandroMachado/BitcoinPaymentURI/pull/6) ([SandroMachado](https://github.com/SandroMachado))

## [1.0.2](https://github.com/SandroMachado/BitcoinPaymentURI/tree/1.0.2) (2016-02-16)
[Full Changelog](https://github.com/SandroMachado/BitcoinPaymentURI/compare/1.0.1...1.0.2)

**Merged pull requests:**

- Update gradle configuration to use a lower java version [\#4](https://github.com/SandroMachado/BitcoinPaymentURI/pull/4) ([SandroMachado](https://github.com/SandroMachado))

## [1.0.1](https://github.com/SandroMachado/BitcoinPaymentURI/tree/1.0.1) (2016-02-16)
[Full Changelog](https://github.com/SandroMachado/BitcoinPaymentURI/compare/1.0.0...1.0.1)

**Merged pull requests:**

- Add package name [\#3](https://github.com/SandroMachado/BitcoinPaymentURI/pull/3) ([SandroMachado](https://github.com/SandroMachado))
- Update README.md [\#2](https://github.com/SandroMachado/BitcoinPaymentURI/pull/2) ([SandroMachado](https://github.com/SandroMachado))
- Add travis [\#1](https://github.com/SandroMachado/BitcoinPaymentURI/pull/1) ([SandroMachado](https://github.com/SandroMachado))

## [1.0.0](https://github.com/SandroMachado/BitcoinPaymentURI/tree/1.0.0) (2016-02-16)


\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

src/test/java/BitcoinPaymentURITest.java
import org.junit.Test;

import static org.junit.Assert.*;

public class BitcoinPaymentURITest {
	
    @Test 
    public void testParseForAddressMethod() {
    	BitcoinPaymentURI bitcoinPaymentURI = BitcoinPaymentURI.parse("bitcoin:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W");
    	
    	assertEquals(bitcoinPaymentURI.getAddress(), "175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W");
    	assertNull(bitcoinPaymentURI.getAmount());
    	assertNull(bitcoinPaymentURI.getLabel());
    	assertNull(bitcoinPaymentURI.getMessage());
    	assertEquals(bitcoinPaymentURI.getParameters().size(), 0);
    }
    
    @Test 
    public void testParseForAddressWithNameMethod() {
    	BitcoinPaymentURI bitcoinPaymentURI = BitcoinPaymentURI.parse("bitcoin:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W?label=Luke-Jr");
    	
    	assertEquals(bitcoinPaymentURI.getAddress(), "175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W");
    	assertNull(bitcoinPaymentURI.getAmount());
    	assertEquals(bitcoinPaymentURI.getLabel(), "Luke-Jr");
    	assertNull(bitcoinPaymentURI.getMessage());
    	assertEquals(bitcoinPaymentURI.getParameters().size(), 0);
    }
    
    @Test 
    public void testParseForAddressWithAmountAndNameMethod() {
    	BitcoinPaymentURI bitcoinPaymentURI = BitcoinPaymentURI.parse("bitcoin:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W?amount=20.3&label=Luke-Jr");
    	
    	assertEquals(bitcoinPaymentURI.getAddress(), "175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W");
    	assertEquals(bitcoinPaymentURI.getAmount(), 20,3);
    	assertEquals(bitcoinPaymentURI.getLabel(), "Luke-Jr");
    	assertNull(bitcoinPaymentURI.getMessage());
    	assertEquals(bitcoinPaymentURI.getParameters().size(), 0);
    }
    
    @Test 
    public void testParseForAddressWithAmountAndNameAndMessageAndRequiredParameterMethod() {
    	BitcoinPaymentURI bitcoinPaymentURI = BitcoinPaymentURI.parse("bitcoin:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W?amount=50&label=Luke-Jr&message=Donation%20for%20project%20xyz");
    	
    	assertEquals(bitcoinPaymentURI.getAddress(), "175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W");
    	assertEquals(bitcoinPaymentURI.getAmount(), Double.valueOf(50));
    	assertEquals(bitcoinPaymentURI.getLabel(), "Luke-Jr");
    	assertEquals(bitcoinPaymentURI.getMessage(), "Donation for project xyz");
    	assertEquals(bitcoinPaymentURI.getParameters().size(), 0);
    }
    
    @Test 
    public void testParseForAddressWithAmountAndNameAndMessageAndParametersMethod() {
    	BitcoinPaymentURI bitcoinPaymentURI = BitcoinPaymentURI.parse("bitcoin:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W?somethingyoudontunderstand=50&somethingelseyoudontget=999&req-app=appname");
    	
    	assertEquals(bitcoinPaymentURI.getAddress(), "175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W");
    	assertNull(bitcoinPaymentURI.getAmount());
    	assertNull(bitcoinPaymentURI.getLabel());
    	assertNull(bitcoinPaymentURI.getMessage());
    	assertEquals(bitcoinPaymentURI.getParameters().size(), 3);
    	assertEquals(bitcoinPaymentURI.getParameters().get("somethingyoudontunderstand").getValue(), "50");
    	assertEquals(bitcoinPaymentURI.getParameters().get("somethingelseyoudontget").getValue(), "999");
    	assertEquals(bitcoinPaymentURI.getParameters().get("app").getValue(), "appname");
    	assertTrue(bitcoinPaymentURI.getParameters().get("app").isRequired());
    }
    
    @Test 
    public void testParseForInvalidAddressesMethod() {
    	BitcoinPaymentURI bitcoinPaymentURI1 = BitcoinPaymentURI.parse("bitcoin:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W?somethingyoudontunderstand=50&somethingelseyoudontget");
    	BitcoinPaymentURI bitcoinPaymentURI2 = BitcoinPaymentURI.parse("bitcoinX:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W?somethingyoudontunderstand=50");
    	BitcoinPaymentURI bitcoinPaymentURI3 = BitcoinPaymentURI.parse("bitcoin175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W?somethingyoudontunderstand=50");
    	BitcoinPaymentURI bitcoinPaymentURI4 = BitcoinPaymentURI.parse("bitcoin:?somethingyoudontunderstand=50");
    	BitcoinPaymentURI bitcoinPaymentURI5 = BitcoinPaymentURI.parse("bitcoin:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W?label");

    	assertNull(bitcoinPaymentURI1);
    	assertNull(bitcoinPaymentURI2);
    	assertNull(bitcoinPaymentURI3);
    	assertNull(bitcoinPaymentURI4);
    	assertNull(bitcoinPaymentURI5);
    }
    
    @Test 
    public void testBuilder() {
    	BitcoinPaymentURI bitcoinPaymentURI = new BitcoinPaymentURI.Builder()
    			.address("175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W")
    			.amount(50.0)
    			.label("Luke-Jr")
    			.message("Donation for project xyz")
    			.parameter("foo", "bar")
    			.requiredParameter("fiz", "biz")
    			.build();
    			    	
    	assertEquals(bitcoinPaymentURI.getAddress(), "175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W");
    	assertEquals(bitcoinPaymentURI.getAmount(), Double.valueOf(50));
    	assertEquals(bitcoinPaymentURI.getLabel(), "Luke-Jr");
    	assertEquals(bitcoinPaymentURI.getMessage(), "Donation for project xyz");
    	assertEquals(bitcoinPaymentURI.getParameters().size(), 2);
    	assertEquals(bitcoinPaymentURI.getParameters().get("foo").getValue(), "bar");
    	assertFalse(bitcoinPaymentURI.getParameters().get("foo").isRequired());
    	assertEquals(bitcoinPaymentURI.getParameters().get("fiz").getValue(), "biz");
    	assertTrue(bitcoinPaymentURI.getParameters().get("fiz").isRequired());
    	assertEquals(bitcoinPaymentURI.getURI(), "bitcoin:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W?message=Donation%20for%20project%20xyz&amount=50.0&req-fiz=biz&foo=bar&label=Luke-Jr");
    }
    
}


src/test/java/com/sandro/bitcoinpaymenturi/test/BitcoinPaymentURITest.java
package com.sandro.bitcoinpaymenturi.test;
import org.junit.Test;

import com.sandro.bitcoinpaymenturi.BitcoinPaymentURI;

import static org.junit.Assert.*;

public class BitcoinPaymentURITest {

    @Test
    public void testParseForAddressMethod() {
    	BitcoinPaymentURI bitcoinPaymentURI = BitcoinPaymentURI.parse("bitcoin:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W");

    	assertEquals(bitcoinPaymentURI.getAddress(), "175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W");
    	assertNull(bitcoinPaymentURI.getAmount());
    	assertNull(bitcoinPaymentURI.getLabel());
    	assertNull(bitcoinPaymentURI.getMessage());
    	assertEquals(bitcoinPaymentURI.getParameters().size(), 0);
    }

    @Test
    public void testParseForAddressWithNameMethod() {
    	BitcoinPaymentURI bitcoinPaymentURI = BitcoinPaymentURI.parse("bitcoin:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W?label=Luke-Jr");

    	assertEquals(bitcoinPaymentURI.getAddress(), "175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W");
    	assertNull(bitcoinPaymentURI.getAmount());
    	assertEquals(bitcoinPaymentURI.getLabel(), "Luke-Jr");
    	assertNull(bitcoinPaymentURI.getMessage());
    	assertEquals(bitcoinPaymentURI.getParameters().size(), 0);
    }

    @Test
    public void testParseForAddressWithAmountAndNameMethod() {
    	BitcoinPaymentURI bitcoinPaymentURI = BitcoinPaymentURI.parse("bitcoin:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W?amount=20.3&label=Luke-Jr");

    	assertEquals(bitcoinPaymentURI.getAddress(), "175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W");
    	assertEquals(bitcoinPaymentURI.getAmount(), 20,3);
    	assertEquals(bitcoinPaymentURI.getLabel(), "Luke-Jr");
    	assertNull(bitcoinPaymentURI.getMessage());
    	assertEquals(bitcoinPaymentURI.getParameters().size(), 0);
    }

    @Test
    public void testParseForAddressWithAmountAndNameAndMessageAndRequiredParameterMethod() {
    	BitcoinPaymentURI bitcoinPaymentURI = BitcoinPaymentURI.parse("bitcoin:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W?amount=50&label=Luke-Jr&message=Donation%20for%20project%20xyz");

    	assertEquals(bitcoinPaymentURI.getAddress(), "175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W");
    	assertEquals(bitcoinPaymentURI.getAmount(), Double.valueOf(50));
    	assertEquals(bitcoinPaymentURI.getLabel(), "Luke-Jr");
    	assertEquals(bitcoinPaymentURI.getMessage(), "Donation for project xyz");
    	assertEquals(bitcoinPaymentURI.getParameters().size(), 0);
    }

    @Test
    public void testParseForAddressWithAmountAndNameAndMessageAndParametersMethod() {
		BitcoinPaymentURI bitcoinPaymentURI = BitcoinPaymentURI.parse("bitcoin:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W?somethingyoudontunderstand=50&somethingelseyoudontget=999&r=https%3A%2F%2Ffoo.com%2Fi%2F7BpFbVsnh5PUisfh&req-app=appname");

    	assertEquals(bitcoinPaymentURI.getAddress(), "175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W");
    	assertNull(bitcoinPaymentURI.getAmount());
    	assertNull(bitcoinPaymentURI.getLabel());
    	assertNull(bitcoinPaymentURI.getMessage());
	assertEquals(bitcoinPaymentURI.getParameters().size(), 4);
    	assertEquals(bitcoinPaymentURI.getParameters().get("somethingyoudontunderstand").getValue(), "50");
    	assertEquals(bitcoinPaymentURI.getParameters().get("somethingelseyoudontget").getValue(), "999");
	assertEquals(bitcoinPaymentURI.getParameters().get("r").getValue(), "https://foo.com/i/7BpFbVsnh5PUisfh");
    	assertEquals(bitcoinPaymentURI.getParameters().get("app").getValue(), "appname");
    	assertTrue(bitcoinPaymentURI.getParameters().get("app").isRequired());
    }

    @Test
    public void testParseForInvalidAddressesMethod() {
    	BitcoinPaymentURI bitcoinPaymentURI1 = BitcoinPaymentURI.parse("bitcoin:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W?somethingyoudontunderstand=50&somethingelseyoudontget");
    	BitcoinPaymentURI bitcoinPaymentURI2 = BitcoinPaymentURI.parse("bitcoinX:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W?somethingyoudontunderstand=50");
    	BitcoinPaymentURI bitcoinPaymentURI3 = BitcoinPaymentURI.parse("bitcoin175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W?somethingyoudontunderstand=50");
    	BitcoinPaymentURI bitcoinPaymentURI4 = BitcoinPaymentURI.parse("bitcoin:?somethingyoudontunderstand=50");
    	BitcoinPaymentURI bitcoinPaymentURI5 = BitcoinPaymentURI.parse("bitcoin:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W?label");

    	assertNull(bitcoinPaymentURI1);
    	assertNull(bitcoinPaymentURI2);
    	assertNull(bitcoinPaymentURI3);
    	assertNull(bitcoinPaymentURI4);
    	assertNull(bitcoinPaymentURI5);
    }

    @Test
    public void testBuilder() {
    	BitcoinPaymentURI bitcoinPaymentURI = new BitcoinPaymentURI.Builder()
    			.address("175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W")
    			.amount(50.0)
    			.label("Luke-Jr")
    			.message("Donation for project xyz")
    			.parameter("foo", "bar")
    			.requiredParameter("fiz", "biz")
    			.build();

    	assertEquals(bitcoinPaymentURI.getAddress(), "175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W");
    	assertEquals(bitcoinPaymentURI.getAmount(), Double.valueOf(50));
    	assertEquals(bitcoinPaymentURI.getLabel(), "Luke-Jr");
    	assertEquals(bitcoinPaymentURI.getMessage(), "Donation for project xyz");
    	assertEquals(bitcoinPaymentURI.getParameters().size(), 2);
    	assertEquals(bitcoinPaymentURI.getParameters().get("foo").getValue(), "bar");
    	assertFalse(bitcoinPaymentURI.getParameters().get("foo").isRequired());
    	assertEquals(bitcoinPaymentURI.getParameters().get("fiz").getValue(), "biz");
    	assertTrue(bitcoinPaymentURI.getParameters().get("fiz").isRequired());
    	assertEquals(bitcoinPaymentURI.getURI(), "bitcoin:175tWpb8K1S7NmH4Zx6rewF9WQrcZv245W?message=Donation%20for%20project%20xyz&amount=50.0&req-fiz=biz&foo=bar&label=Luke-Jr");
    }

}


src/main/java/com/sandro/bitcoinpaymenturi/BitcoinPaymentURI.java
package com.sandro.bitcoinpaymenturi;
import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.sandro.bitcoinpaymenturi.model.Parameter;

/**
 * Java library to handle Bitcoin payment URI.
 * This library is based on the specification at the BIP 21.
 *
 * The BIT is available at: https://github.com/bitcoin/bips/blob/master/bip-0021.mediawiki
 */

public class BitcoinPaymentURI {

	private static final String SCHEME = "bitcoin:";
	private static final String PARAMETER_AMOUNT = "amount";
	private static final String PARAMETER_LABEL = "label";
	private static final String PARAMETER_MESSAGE = "message";

	private final String address;
	private final HashMap<String, Parameter> parameters;

	private BitcoinPaymentURI(Builder builder) {
		this.address = builder.address;

		parameters = new HashMap<String, Parameter>();

		if (builder.amount != null) {
			parameters.put(PARAMETER_AMOUNT, new Parameter(String.valueOf(builder.amount), false));
		}

		if (builder.label != null) {
			parameters.put(PARAMETER_LABEL, new Parameter(builder.label, false));
		}

		if (builder.message != null) {
			parameters.put(PARAMETER_MESSAGE, new Parameter(builder.message, false));
		}

		if (builder.otherParameters != null) {
			parameters.putAll(builder.otherParameters);
		}
	}

	/**
	 * Gets the URI Bitcoin address.
	 *
	 * @return the URI Bitcoin address.
	 */

	public String getAddress() {
		return address;
	}

	/**
	 * Gets the URI amount.
	 *
	 * @return the URI amount.
	 */

	public Double getAmount() {
		if (parameters.get(PARAMETER_AMOUNT) == null) {
			return null;
		}

		return Double.valueOf(parameters.get(PARAMETER_AMOUNT).getValue());
	}

	/**
	 * Gets the URI label.
	 *
	 * @return the URI label.
	 */

	public String getLabel() {
		if (parameters.get(PARAMETER_LABEL) == null) {
			return null;
		}

		return parameters.get(PARAMETER_LABEL).getValue();
	}

	/**
	 * Gets the URI message.
	 *
	 * @return the URI message.
	 */

	public String getMessage() {
		if (parameters.get(PARAMETER_MESSAGE) == null) {
			return null;
		}

		return parameters.get(PARAMETER_MESSAGE).getValue();
	}

	/**
	 * Gets the URI parameters.
	 *
	 * @return the URI parameters.
	 */

	public HashMap<String, Parameter> getParameters() {
		HashMap<String, Parameter> filteredParameters = new HashMap<String, Parameter>(parameters);

		filteredParameters.remove(PARAMETER_AMOUNT);
		filteredParameters.remove(PARAMETER_LABEL);
		filteredParameters.remove(PARAMETER_MESSAGE);

		return filteredParameters;
	}

	/**
	 * Gets the URI.
	 *
	 * @return a string with the URI. This string can be used to make a Bitcoin payment.
	 */

	public String getURI() {
		String queryParameters = null;
		try {
			for (Map.Entry<String, Parameter> entry : parameters.entrySet()) {
				if (queryParameters == null) {
					if (entry.getValue().isRequired()) {
						queryParameters = String.format("req-%s=%s", URLEncoder.encode(entry.getKey(), "UTF-8").replace("+", "%20"), URLEncoder.encode(entry.getValue().getValue(), "UTF-8").replace("+", "%20"));

						continue;
					}

					queryParameters = String.format("%s=%s", URLEncoder.encode(entry.getKey(), "UTF-8").replace("+", "%20"), URLEncoder.encode(entry.getValue().getValue(), "UTF-8").replace("+", "%20"));

					continue;
				}

				if (entry.getValue().isRequired()) {
					queryParameters = String.format("%s&req-%s=%s", queryParameters, URLEncoder.encode(entry.getKey(), "UTF-8").replace("+", "%20"), URLEncoder.encode(entry.getValue().getValue(), "UTF-8").replace("+", "%20"));

					continue;
				}

				queryParameters = String.format("%s&%s=%s", queryParameters, URLEncoder.encode(entry.getKey(), "UTF-8").replace("+", "%20"), URLEncoder.encode(entry.getValue().getValue(), "UTF-8").replace("+", "%20"));
			}

		} catch (UnsupportedEncodingException e) {
			e.printStackTrace();

			return null;
		}

		return String.format("%s%s%s", SCHEME, getAddress(), queryParameters == null ? "" : String.format("?%s", queryParameters));
	}

	/**
	 * Parses a string to a Bitcoin payment URI.
	 *
	 * @param string The string to be parsed.
	 *
	 * @return a Bitcoin payment URI if the URI is valid, or null for an invalid string.
	 */

	public static BitcoinPaymentURI parse(String string) {
		try {
			string = URLDecoder.decode(string,  "UTF-8");
		} catch (UnsupportedEncodingException e) {
			e.printStackTrace();

			return null;
		}

		if (string == null) {
			return null;
		}

		if (string.isEmpty()) {
			return null;
		}
		
		if (!string.toLowerCase().startsWith(SCHEME)) {
			return null;
		}

		String bitcoinPaymentURIWithoutScheme = string.replaceFirst(SCHEME, "");
        ArrayList<String> bitcoinPaymentURIElements = new ArrayList<>(Arrays.asList(bitcoinPaymentURIWithoutScheme.split("\\?")));

        if (bitcoinPaymentURIElements.size() != 1 && bitcoinPaymentURIElements.size() != 2) {
        	return null;
        }

        if (bitcoinPaymentURIElements.get(0).length() == 0) {
        	return null;
        }

        if (bitcoinPaymentURIElements.size() == 1) {
        	return new Builder().address(bitcoinPaymentURIElements.get(0)).build();
        }

        List<String> queryParametersList = Arrays.asList(bitcoinPaymentURIElements.get(1).split("&"));

        if (queryParametersList.isEmpty()) {
        	return new Builder().address(bitcoinPaymentURIElements.get(0)).build();
        }

        HashMap<String, String> queryParametersFiltered = new HashMap<String, String>();

        for (String query : queryParametersList) {
        	String[] queryParameter = query.split("=");

        	try {
            	queryParametersFiltered.put(queryParameter[0], queryParameter[1]);
        	}catch(ArrayIndexOutOfBoundsException exception) {
        		exception.printStackTrace();

        		return null;
        	}
        }

        Builder bitcoinPaymentURIBuilder = new Builder().address(bitcoinPaymentURIElements.get(0));

        if (queryParametersFiltered.containsKey(PARAMETER_AMOUNT)) {
        	bitcoinPaymentURIBuilder.amount(Double.valueOf(queryParametersFiltered.get(PARAMETER_AMOUNT)));

        	queryParametersFiltered.remove(PARAMETER_AMOUNT);
        }

        if (queryParametersFiltered.containsKey(PARAMETER_LABEL)) {
        	bitcoinPaymentURIBuilder.label(queryParametersFiltered.get(PARAMETER_LABEL));

        	queryParametersFiltered.remove(PARAMETER_LABEL);
        }

        if (queryParametersFiltered.containsKey(PARAMETER_MESSAGE)) {
        	bitcoinPaymentURIBuilder.message(queryParametersFiltered.get(PARAMETER_MESSAGE));

        	queryParametersFiltered.remove(PARAMETER_MESSAGE);
        }

		for (Map.Entry<String, String> entry : queryParametersFiltered.entrySet()) {
			bitcoinPaymentURIBuilder.parameter(entry.getKey(), entry.getValue());
		}

		return bitcoinPaymentURIBuilder.build();
	}

	public static class Builder{

		private String address;
		private Double amount;
		private String label;
		private String message;
		private HashMap<String, Parameter> otherParameters;

		/**
		 * Returns a builder for the Bitcoin payment URI.
		 */

		public Builder() {
		}

		/**
		 * Adds the address to the builder.
		 *
		 * @param address The address.
		 *
		 * @return the builder with the address.
		 */

		public Builder address(String address) {
			this.address = address;

			return this;
		}

		/**
		 * Adds the amount to the builder.
		 *
		 * @param amount The amount.
		 *
		 * @return the builder with the amount.
		 */

		public Builder amount(Double amount) {
			this.amount = amount;

			return this;
		}

		/**
		 * Adds the label to the builder.
		 *
		 * @param label The label.
		 *
		 * @return the builder with the label.
		 */

		public Builder label(String label) {
			this.label = label;

			return this;
		}

		/**
		 * Adds the message to the builder.
		 *
		 * @param message The message.
		 *
		 * @return the builder with the message.
		 */

		public Builder message(String message) {
			this.message = message;

			return this;
		}

		/**
		 * Adds a parameter to the builder.
		 *
		 * @param key The parameter.
		 * @param value The value.
		 *
		 * @return the builder with the parameter.
		 */

		public Builder parameter(String key, String value) {
			if (otherParameters == null) {
				otherParameters = new HashMap<String, Parameter>();
			}

			if (key.startsWith("req-")) {
				otherParameters.put(key.replace("req-", ""), new Parameter(value, true));

				return this;
			}

			otherParameters.put(key, new Parameter(value, false));

			return this;
		}

		/**
		 * Adds a required to the builder.
		 *
		 * @param key The key.
		 * @param value The value.
		 *
		 * @return the builder with the parameter.
		 */

		public Builder requiredParameter(String key, String value) {
			if (otherParameters == null) {
				otherParameters = new HashMap<String, Parameter>();
			}

			otherParameters.put(key, new Parameter(value, true));

			return this;
		}

		/**
		 * Builds a Bitcoin payment URI.
		 *
		 * @return a Bitcoin payment URI.
		 */

		public BitcoinPaymentURI build() {
			return new BitcoinPaymentURI(this);
		}

	}

}
